### ISSUE TITLE
Make it possible to create truely dynamic queries in Concepts:

### ISSUE THREAD

**Issue by agharbeia** [SemanticMediaWiki/SemanticMediaWiki#3504]
This is a feature suggestion:

It would useful to be able to pass at least values of the query restriction criteria (values of queried properties) to a defined Concept.

For example one could create a Concept:Author_in_a_decade to query the books authored by a given author in a given decade, and then be able to use it to query books authored by another author in another decade, without having to create a concept for for each case.

With a simple query such as this maybe an inline query would be feasible, but more complex cases are possible. Additionally, the ability to link to the Concept page, while passing parameters to it, is very useful; even while assuming the formatting parameters of the Concept will have to be fixed.

Or will this make things unjustifiably complex?

See also the related #3503

**Comment by kghbln** [comment]
Thanks for suggesting. Please allow time to comment.

**Comment by mwjames** [comment]
> It would useful to be able to pass at least values of the query restriction criteria (values of queried properties) to a defined Concept. 

Concepts are static in terms of the defined query they represent with that being “… dynamic ...” is not an objective of a concept instance. A concept provides an instance representation of subjects (and can be cached) that are determined by a predefined query.

As outlined in the documentation, concepts can be seen as “dynamic categories” [0] where dynamic is a reference to an individual query to collect a list of subjects that satisfy the condition of a query defined by a concept.

`{{#concept: ...}}` defines a concept and the embedding concept page is nothing more than a representation of that state not the concept itself.

There are ways of making a concept query a bit more flexible in a temporal context such as [1] but that is also predetermined at the time of its execution and does not allow for injection of user “dynamic” parameters during the execution.

In short, concepts are not expected to be modified when being part of another query or reference thereof.

> For example one could create a Concept:Author_in_a_decade to query the books authored by a given author in a given decade, and then be able to use it to query books authored by another author in another decade, without having to create a concept for for each case. 

In your described use case, you would create a `Concept:Books_in_decade` and apply the filtering of authors individualy such as `[[Concept:Books_in_decade]] && [[Author::Foo]] && [[Author::Bar]]`, so there is no need to modify a concept or change its behavioural pattern.

> With a simple query such as this maybe an inline query would be feasible, but more complex cases are possible. Additionally, the ability to link to the Concept page, while passing parameters to it,  

This is unrelated to a concept, if you need individual filtering of values to match a specific constraint or content then this should be part of the query and not be part of a concept.

[0] https://www.semantic-mediawiki.org/wiki/Help:Concepts
[1] https://wikiapiary.com/wiki/Concept:Birthday_today

**Comment by kghbln** [comment]
Thanks @mwjames for elaborating. I am closing this since I believe the suggestions was considered.

**Comment by agharbeia** [comment]
I probably was not clear at all. Let me try again, with clearer examples:
> > For example one could create a Concept:Author_in_a_decade to query the books authored by a given author in a given decade, and then be able to use it to query books authored by another author in another decade, without having to create a concept for for each case.
> 
> In your described use case, you would create a `Concept:Books_in_decade` and apply the filtering of authors individualy such as `[[Concept:Books_in_decade]] && [[Author::Foo]] && [[Author::Bar]]`, so there is no need to modify a concept or change its behavioural pattern.

But according to your example I will have to create a concept for each decade I would like to feature, before further filtering its result by another query that introduces the desired specific author restriction.

What I'm hoping to acheive, however, is create a skeleton of a query that has some basic setting such as specific categories and properties to filter on, but which gets the desired values/range of these properties from outside; by passing them as parameters, preferably as a wikilinks.

That way it would be possible to create a template that links to a query/Concept: based on contextual information and without having to write a query for each possible case.

**example**:
template:book_author_decade:
```
* [[concept:Books_Author_Decade/a={{{a|{{{1|}}}}}}/d={{{d|{{{2|}}}}}} | Books by {{{a|{{{1|}}}}}} in {{{d|{{{2|}}}}}}]]
```
By invoking the template as {{book_author_decade|Heinlein|1930}} one would get:
* [[concept:Books_Author_Decade/d=1930/a=Heinlein|Books in 1930 by Heinline]]

and similarly, with the appropriate parameter values:
* [[concept:Books_Author_Decade/d=1940/a=Heinlein|Book in 1940s by Heinlein]]
* [[concept:Books_Author_Decade/d=1940/a=Asimov|Books in 1940s by Asimov]]
* Books by x in yyyy
And the concept would display a dataset based on the passed criteria:
```
{{#concept:
 [[Category:Books]]
 [[Has Author::{{{a|{{{1|}}}}}}]]
 [[Has year published::> {{{d|{{{2|}}}}}}]]
 [[Has year published::< {{#sub:{{{d|{{{2|}}}}}} | 0 | 3}}9]] 
}}
```
